#5475 Allow read device_id from the command line args to be used in o…

sfbirdf opened the pull request:
…penni_3d_concave_hull.cpp.

mvieth commented:
Thanks, looks good.
Can you also check which of the following files need this fix:
```
apps/3d_rec_framework/src/tools/openni_frame_source.cpp
apps/src/openni_tracking.cpp
apps/src/openni_boundary_estimation.cpp
apps/src/openni_klt.cpp
apps/src/openni_organized_edge_detection.cpp
apps/src/openni_mls_smoothing.cpp
apps/src/openni_uniform_sampling.cpp
apps/src/openni_ii_normal_estimation.cpp
apps/src/openni_grab_images.cpp
apps/src/openni_voxel_grid.cpp
apps/src/openni_feature_persistence.cpp
apps/src/openni_3d_concave_hull.cpp
apps/src/openni_fast_mesh.cpp
apps/src/openni_planar_convex_hull.cpp
apps/src/face_detection/openni_frame_source.cpp
apps/src/openni_planar_segmentation.cpp
apps/src/openni_color_filter.cpp
apps/src/openni_3d_convex_hull.cpp
apps/src/openni_mobile_server.cpp
```

sfbirdf commented:
[PR - #5475.pdf](https://github.com/PointCloudLibrary/pcl/files/9815389/PR.-.5475.pdf)

I just combed through all the files, here are my detailed findings - I can also start working on them if you'd like, just say the word. It'll probably take some time for me since I'm still a bit afraid of touching open source code, but I'm definitely willing and interested :D

mvieth commented:
Wow, that is really detailed, thank you for the analysis. Would be great if you could also fix the other files, you can add the extra changes to this pull request. Regarding openni_mobile_server.cpp: do you think it makes sense to just use `argv[1]` as the device id unless that is `-port`, `-leaf`, `-h`, or `--help`, or `argc` is 1?

sfbirdf commented:
Hey, sorry for the late response :)

I fixed the rest of the files - there are several things to mention:
- in `openni_voxel_grid.cpp` we don't return after `usage(argv)`, I added a return statement (like all the other files). I hope that would be okay.
- regarding `openni_mobile_server.cpp` - I agree! I modified `usage` to fit the new format for the command line args and added a bit to the parsing, just as how it's implemented in other files.
- this is unrelated, but I noticed that sometimes --h and --help would work, sometimes just --h.

larshg commented:
You can go to the build server at see whats not in its liking : https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=22161&view=logs&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b

It seems to be some unneeded whitespace, and probably missing braces at the return 1, which you placed under an if. But to gave multiple lines under an if you need {} braces.

larshg commented:
And no need to be sorry for daily reply. We all need to sleep and eventually there is time difference between people as well 👍

larshg commented:
Some more https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=22165&view=logs&j=84ba2080-ade3-55e3-f209-94ce73f9d406&t=9838d42f-a715-5615-242f-7e73bf03078b

sfbirdf commented:
Thank you for the tip! It just passed the formatting check :)

larshg commented:
Yeah, nice 👌

larshg commented:
Now there is compile errors though: see https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=22167&view=logs&j=149f4e41-9368-548e-d6d4-ff21d16f0ace&t=30234468-9402-53e3-1a03-9eb9da598fc6&l=1632

You can find the builds under "checks" which links to the azure builds.

larshg commented:
I would suggest to compile locally first. But if you only build on linux, then you can push at test windows (and other builds) on the build servers.

Yes, to ensure to build everything, additional flags need to be set either using cmake arguments or the cmake gui(need to run multiple times, ie. after checkbox changes)

No need to apologise for mistakes. Thats why the CI's are there 😄

sfbirdf commented:
Hey! All checks seem to have passed, I'll be happy to have them reviewed.

larshg commented on apps/src/openni_mobile_server.cpp:
```diff
@@ -230,19 +230,23 @@ usage(char** argv)
   int port = 11111;
   float leaf_x = 0.01f, leaf_y = 0.01f, leaf_z = 0.01f;
-  std::string device_id;
+  std::string device_id = arg;
```
```suggestion
```

larshg replied:
Remove this variable and save the id directly in the device_id field.

larshg commented on apps/src/openni_mobile_server.cpp:
```diff
@@ -230,19 +230,23 @@ usage(char** argv)
 {
+  std::string arg;
+  if (argc > 1)
+    arg = std::string(argv[1]);
```
```suggestion
    device_id = std::string(argv[1]);
```

larshg commented on apps/src/openni_mobile_server.cpp:
```diff
@@ -230,19 +230,23 @@ usage(char** argv)
 int
 main(int argc, char** argv)
 {
+  std::string arg;
```
```suggestion
  std::string device_id;
```

larshg commented:
Generally I would rename arg to device_id, since it is more clear what it is. Arg doesn't really say much.

sfbirdf commented:
Hey! So, keep the arg there, but change the if statements to `pcl::console::find_argument(argc, argv, "-h")`, right?

sfbirdf commented:
Hey again, I just fixed the mobile server. Maybe it is a good idea to merge/close this and open up a new issue? Now that I am quite familiar with the constructors and parsing functions, I'd be glad to work on reworking parsing/making a unified argument parsing structure for the app source files, as I saw quite a few inconsistencies here and there (an example would be the temporary variables and also -h and --help).

larshg commented:
Yes, we can do that - looks good 👍

sfbirdf commented:
Thank you! This commit seems to build; please notify me here if we're ready to move on with that :))

mvieth commented on apps/src/openni_feature_persistence.cpp:
```diff
@@ -266,6 +266,10 @@ main(int argc, char** argv)
             << "Use \"-h\" to get more info about the available options.\n";
 
+  std::string arg;
+  if (argc > 1)
```
Might be best to also test that the first arg isn't one of the other options (`-octree_leaf_size`, `-normal_search_radius`, etc) so that giving the device id on the command line is still optional. Checking whether the arg begins with a `-` should be enough.
Same for openni_mobile_server.cpp and openni_voxel_grid.cpp

sfbirdf commented:
Hi! I made the changes, but it won't build for MSVC x64 for some reason. . -> https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=22188&view=logs&j=b7dd6ba8-1ad9-5018-2280-3f36c72607c9&t=2fd7de77-c8c6-5479-ec09-1c18859bcc95&l=261

The problem seems to be unresolved external symbols at open_3d.cpp and voxel_grid.cpp, could you please take a look at it?

sfbirdf commented:
Apparently my previously working build won't build for MSVC x64 either.

mvieth reviewed (approved):
Thank you! Looks good to me.
The failing MSVC x64 build is unrelated to your changes, so we can ignore that. As long as the other builds succeed, this is ready to merge from my side.

sfbirdf commented:
Hi! Everything seems to have built, could we merge? :D
